core/src/main/scala/org/apache/spark/scheduler/ExecutorBlacklistTracker.scala

```diff
@@ -0,0 +1,175 @@
+
+  def start(): Unit = {
+    val scheduleTask = new Runnable() {
+      override def run(): Unit = {
+        Utils.logUncaughtExceptions(expireTimeoutExecutorBlacklist())
+      }
+    }
+    executor.scheduleAtFixedRate(scheduleTask, 0L, 60, TimeUnit.SECONDS)
+  }
+
+  def stop(): Unit = {
+    executor.shutdown()
+    executor.awaitTermination(10, TimeUnit.SECONDS)
+  }
+
+  def setClock(newClock: Clock): Unit = {
+    clock = newClock
+  }
+
+  def getExecutorBlacklist: Set[String] = synchronized {
+    executorIdToTaskFailures.filter(_._2.isBlackListed).keys.toSet
+  }
+
+  override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = synchronized {
+    taskEnd.reason match {
+      case _: FetchFailed | _: ExceptionFailure | TaskResultLost |
+          _: ExecutorLostFailure | UnknownReason =>
+        val failureStatus = executorIdToTaskFailures.getOrElseUpdate(taskEnd.taskInfo.executorId,
+          new ExecutorFailureStatus)
+        failureStatus.numFailures += 1
+        failureStatus.updatedTime = clock.getTimeMillis()
+
+        // Update the executor blacklist
+        updateExecutorBlacklist()
+      case _ => Unit
+    }
+  }
+
+  override def onExecutorAdded(executorAdded: SparkListenerExecutorAdded): Unit = {
+    numExecutorsRegistered += 1
+  }
+
+  override def onExecutorRemoved(executorRemoved: SparkListenerExecutorRemoved
+      ): Unit = synchronized {
+    numExecutorsRegistered -= 1
+    executorIdToTaskFailures -= executorRemoved.executorId
+  }
+
+  private def updateExecutorBlacklist(): Unit = {
+    // Filter out the executor Ids where task failure number is larger than
+    // executorFaultThreshold and not blacklisted
+    val failedExecutors = executorIdToTaskFailures.filter { case(_, e) =>
+      e.numFailures >= executorFaultThreshold && !e.isBlackListed
+    }
+
+    val blacklistedExecutorNum = executorIdToTaskFailures.filter(_._2.isBlackListed).size
+
+    if (failedExecutors.nonEmpty) {
+      val avgNumFailed = executorIdToTaskFailures.values.map(_.numFailures).sum.toDouble /
+        numExecutorsRegistered
+      for ((executorId, failureStatus) <- failedExecutors) {
+        // If the number of failure task is more than average blacklist threshold of average
+        // failed number and current executor blacklist is less than the max fraction of number
+        // executors
+        if ((failureStatus.numFailures.toDouble > avgNumFailed * (1 + avgBlacklistThreshold)) &&
+          (blacklistedExecutorNum.toDouble < numExecutorsRegistered * maxBlacklistFraction)) {
+          failureStatus.isBlackListed = true
+        }
+      }
+    }
+  }
+
+  private def expireTimeoutExecutorBlacklist(): Unit = synchronized {
+    val now = clock.getTimeMillis()
+
+    executorIdToTaskFailures.foreach { case (id, failureStatus) =>
+      if ((now - failureStatus.updatedTime) > executorFaultTimeoutWindowInMinutes * 60 * 1000
+        && failureStatus.isBlackListed) {
+        failureStatus.isBlackListed = false
```

Comment:
Yes, that's what I thought about, I'm not sure which way is a better, I checked the MR code, where they don't clean the `numFailures` as I know. So maybe this should be carefully rethink about, thanks for your pointing out.